radixdlt-engine/src/main/java/com/radixdlt/application/system/construction/EpochConstructionState.java

```diff
@@ -0,0 +1,334 @@
+  }
+
+  private static SubstateIndex<ExitingStake> exittingStakeIndex(EpochData closingEpoch) {
+    var unlockedStateIndex =
+        ByteBuffer.allocate(2 + Long.BYTES)
+            .put(SubstateTypeId.EXITING_STAKE.id())
+            .put((byte) 0)
+            .putLong(closingEpoch.epoch() + 1)
+            .array();
+
+    return SubstateIndex.create(unlockedStateIndex, ExitingStake.class);
+  }
+
+  private static TreeSet<ExitingStake> collectExittingStake(Iterator<ExitingStake> iterator) {
+    var exit =
+        new TreeSet<>(
+            Comparator.comparing(ExitingStake::dataKey, UnsignedBytes.lexicographicalComparator()));
+    iterator.forEachRemaining(exit::add);
+    return exit;
+  }
+
+  public void processEmission() {
+    txBuilder
+        .shutdownAll(ValidatorBFTData.class, EpochConstructionState::collectToMap)
+        .forEach(this::calculateEmission);
+  }
+
+  private void calculateEmission(ECPublicKey publicKey, ValidatorBFTData bftData) {
+    if (bftData.completedProposals() + bftData.missedProposals() == 0) {
+      return;
+    }
+
+    var percentageCompleted =
+        bftData.completedProposals()
+            * 10000
+            / (bftData.completedProposals() + bftData.missedProposals());
+
+    // Didn't pass threshold, no rewards!
+    if (percentageCompleted < constructor.minimumCompletedProposalsPercentage()) {
+      return;
+    }
+
+    var nodeRewards =
+        constructor.rewardsPerProposal().multiply(UInt256.from(bftData.completedProposals()));
+    if (nodeRewards.isZero()) {
+      return;
+    }
+
+    var validatorStakeData = stakeData(publicKey);
+    int rakePercentage = validatorStakeData.getRakePercentage();
+    final UInt256 rakedEmissions;
+
+    if (rakePercentage != 0) {
+      var rake = nodeRewards.multiply(UInt256.from(rakePercentage)).divide(UInt256.from(RAKE_MAX));
+      var validatorOwner = validatorStakeData.getOwnerAddr();
+
+      var initStake = createStakeMap();
+      initStake.put(validatorOwner, rake);
+
+      preparingStake.put(publicKey, initStake);
+      rakedEmissions = nodeRewards.subtract(rake);
+    } else {
+      rakedEmissions = nodeRewards;
+    }
+    validatorStakeData.addEmission(rakedEmissions);
+  }
+
+  public void processPreparedUnstake() {
+    txBuilder
+        .shutdownAll(PreparedUnstakeOwnership.class, EpochConstructionState::collectUnstake)
+        .forEach(
+            (publicKey, unstakes) -> {
+              var curValidator = stakeData(publicKey);
+
+              unstakes.forEach(
+                  (owner, amount) -> {
+                    var epochUnlocked =
+                        closingEpoch().epoch() + 1 + constructor.unstakingEpochDelay();
+                    txBuilder().up(curValidator.unstakeOwnership(owner, amount, epochUnlocked));
+                  });
+
+              validatorsToUpdate().put(publicKey, curValidator);
+            });
+  }
+
+  private static TreeMap<ECPublicKey, TreeMap<REAddr, UInt256>> collectUnstake(
```

Comment:
Could we use Objects to represent those two TreeMap's?

Reply:
I've tried to do this, and it requires much deeper rework.